ports/esp32/mpthreadport.c

```diff
@@ -221,32 +212,17 @@ void mp_thread_mutex_unlock(mp_thread_mutex_t *mutex) {
 }
 
 void mp_thread_deinit(void) {
-    for (;;) {
-        // Find a task to delete
-        TaskHandle_t id = NULL;
-        mp_thread_mutex_lock(&thread_mutex, 1);
-        for (mp_thread_t *th = thread; th != NULL; th = th->next) {
-            // Don't delete the current task
-            if (th->id != xTaskGetCurrentTaskHandle()) {
-                id = th->id;
-                break;
-            }
-        }
-        mp_thread_mutex_unlock(&thread_mutex);
+    assert(thread_entry0.id == xTaskGetCurrentTaskHandle());
 
-        if (id == NULL) {
-            // No tasks left to delete
-            break;
-        } else {
-            // Call FreeRTOS to delete the task (it will call vTaskPreDeletionHook)
-            vTaskDelete(id);
+    // Delete all tasks except the main one.
+    mp_thread_mutex_lock(&thread_mutex, 1);
+    for (mp_thread_t *th = thread; th != NULL; th = th->next) {
+        if (th != &thread_entry0) {
+            vTaskDelete(th->id);
         }
     }
-}
-
-#else
-
-void vTaskPreDeletionHook(void *tcb) {
+    thread = &thread_entry0;
```

Comment:
I think `thread_entry0.next == NULL` is an invariant (because new threads always pushed to the head of the list), but it might be worth either asserting that condition here or setting `thread->next = NULL` in case something changes in the future.

Reply:
Added an assertion.